plat/common/include/uk/plat/common/memory.h

```diff
@@ -33,6 +33,108 @@
 #ifndef __PLAT_CMN_MEMORY_H__
 #define __PLAT_CMN_MEMORY_H__
 
+#include <uk/config.h>
+#include <uk/assert.h>
+#include <uk/essentials.h>
+#include <uk/arch/types.h>
+#include <uk/plat/memory.h>
+
+#include <errno.h>
+#include <string.h>
+
+struct ukplat_memregion_list {
+	/** Maximum regions in the list */
+	__u32 capacity;
+	/** Current number of regions in the list */
+	__u32 count;
+	/** Array of memory regions */
+	struct ukplat_memregion_desc mrds[];
+} __packed __align(__SIZEOF_LONG__);
+
+/**
+ * Initializes the fields of an empty memory region list.
+ *
+ * @param list
+ *   The list to initialize
+ * @param capacity
+ *   The number of memory region descriptors that the list can hold. This
+ *   number must be set according to the amount of allocated memory.
+ */
+static inline void
+ukplat_memregion_list_init(struct ukplat_memregion_list *list, __u32 capacity)
+{
+	UK_ASSERT(capacity >= 1);
+
+	list->capacity = capacity;
+	list->count    = 0;
+}
+
+/**
+ * Inserts the region into the memory region list. The list is kept in
+ * ascending order of physical base addresses. Order is only preserved among
+ * ranges that do not overlap.
+ *
+ * @param list
+ *   The memory region list to insert the range into
+ * @param mrd
+ *   The memory range to insert
+ *
+ * @return
+ *   The index of the element added on success, an negative errno otherwise
+ */
+static inline int
+ukplat_memregion_list_insert(struct ukplat_memregion_list *list,
+			     const struct ukplat_memregion_desc *mrd)
+{
+	struct ukplat_memregion_desc *p;
+	__u32 i;
+
+	if (unlikely(list->count == list->capacity))
+		return -ENOMEM;
+
+	/* We start from the tail as it is more likely that we add memory
+	 * regions in semi-sorted order.
+	 */
+	p = &list->mrds[list->count];
+	for (i = list->count; i > 0; i--) {
+		--p;
+
+		if (p->pbase + p->len <= mrd->pbase) {
+			++p;
+			break;
+		}
+	}
+
+	memmove(p + 1, p, sizeof(*p) * (list->count - i));
+
+	*p = *mrd;
+	list->count++;
+	return i;
+}
+
+/**
+ * Delete the specified region from the memory list.
+ *
+ * @param list
+ *   The memory region list from which to delete the region
+ * @param idx
+ *   The index of the region to delete
+ */
+static inline void
+ukplat_memregion_list_delete(struct ukplat_memregion_list *list, __u32 idx)
```

Comment:
Would it be better to have this function receive a `ukplat_memregion_desc  *` as a second argument instead of an index? I believe this would make it more uniform with the insert function as well as easier to use inside a `foreach` block, since you do not have an index there. If the index is required for sanity checks perhaps something as easy as
```
static inline __u32
ukplat_memregion_list_get_idx(struct ukplat_memregion_list *list,
			      const struct ukplat_memregion_desc *mrd)
{
	return mrd - list->mrds;
}
```
could be used inside the delete function.

Reply:
Actually, nevermind about the `foreach` block, this might mess up the next `ukplat_memregion_find_next` since it iterates ascendingly.

Reply:
Using the pointer would suggest to also change `ukplat_memregion_get()` which uses an index, too. And changing that leads to changes in quite a lot of places. Furthermore, using a ptr in `ukplat_memregion_list_delete()` etc. would also make it possible to supply memregion objects from a different list, which is hard to check. So I'll keep the index, if you don't mind.